### ISSUE TITLE
kernel panic in vm_console_init on RaspberryPi 3B

### ISSUE THREAD

**Issue by aananthcn** [minosproject/minos#12]
I followed the instruction in the link below:
https://github.com/minosproject/minos/blob/master/Documents/009-Test_Minos_on_Raspberry_3B.md

But the Linux kernel during the init is throwing "kernel panic", please find the logs attached:
[kernel-panic.log](https://github.com/minosproject/minos/files/5009897/kernel-panic.log)

The cause seems to be NULL pointer dereferencing happening in vm_console_init(), which is caused by the NULL value returned by create_vm_debug_console() function. The failure point within create_vm_debug_console() seems to be due to the id value returned by minos_hvc0() call is not meeting the expected value.

        id = minos_hvc0(HVC_DC_GET_STAT);
        if ((id & 0xffff0000) != 0xabcd0000) {
                return NULL;
        }

On further discussion over email with Le Min, I added few pr_notice("Debug: *") calls in create_dconsole() function defined in virt/debug_console.c in minos-hypervisor, but those debug prints were not seen on console.

Please let me know the next steps to resolve this issue. Am I missing something?

**Comment by lemin9538** [comment]
The information of debug console is missing in the dts file for RPI3b, please add below lines in the dts file for 3b.

diff --git a/dtbs/bcm2837-rpi-3-b-plus.dts b/dtbs/bcm2837-rpi-3-b-plus.dts
index 1d7199f..0967a7b 100644
--- a/dtbs/bcm2837-rpi-3-b-plus.dts
+++ b/dtbs/bcm2837-rpi-3-b-plus.dts
@@ -25,6 +25,15 @@
                        memory = <0x0 0x00000000 0x0 0x28000000>;
                        cmdline = "8250.nr_uarts=1 dwc_otg.lpm_enable=0 net.ifnames=0 earlycon=uart8250,mmio32,0x3f215040 console=ttyS0,115200 cma=64M root=/dev/mmcblk0p2 rw rootfstype=ext4 elevator=deadline fsck.repair=yes rootwait noinitrd";
                };
+
+               vm0_bdi {
+                       vm_console_vm0 {
+                               virtual_device;
+                               vc-dynamic-res;
+                               compatible = "minos,vm_console";
+                       };
+               };
+
        };
 
        soc {

**Comment by lemin9538** [comment]
But I see kernel do not use the debug console as the default tty. So I think this issue is a debug console driver bug in linux.

**Comment by aananthcn** [comment]
I added the patch mentioned in the previous comment and I still see that `create_vm_debug_console()` is failing and returning NULL. I am using `bootargs` as provided in _Note (2)_ below, should I try passing to some specific console device nodes?


Note(s):
(1) I am using 64-bit Raspbian image / rootfs as the base image.
(2) The **bootargs** that I have configured is **console=serial0,115200 console=tty1 root=/dev/mmcblk0p2 rootfstype=ext4 elevator=deadline fsck.repair=yes rootwait quiet splash plymouth.ignore-serialconsoles**

**Comment by lemin9538** [comment]
Please try below patch in kernel driver.

`diff --git a/drivers/minos/vmbox/vmbox_console.c b/drivers/minos/vmbox/vmbox_console.c
index 9fce450cc964..b7ce20f258ff 100644
--- a/drivers/minos/vmbox/vmbox_console.c
+++ b/drivers/minos/vmbox/vmbox_console.c
@@ -623,6 +623,9 @@ static int __init vm_init_debug_console(void)
        if (!vc)
                vc = create_vm_debug_console();
 
+       if (!vc)
+               return -ENODEV;
+
        pr_info("register vm debug console\n");
        hvc_instantiate(VMBOX_HVC_COOLIE + vc->id,
                        vc->id, &vm_debug_hvc_ops);
`

But it is very strange, after you add the previous patch in dts, have you replace the dtb file ?

**Comment by aananthcn** [comment]
With your patch for kernel driver (above) the Linux boots 👍, I can get Linux login prompt and can see 2 vCPUs running on VM0. Thanks. 

But, I could not get the hypervisor console 👎 :-). I could see following error prints in boot log.
```
[    0.012873] vmbox hvc console init for backend
[    0.017450] can not find the hvc console device
```
The complete bootlog is attached here: 
[minicom.log](https://github.com/minosproject/minos/files/5011755/minicom.log)

I can confirm that I have applied the previous patch and copied to the SD Card. I have double confirmed it by adding my name (_later deleted it_) in the cmdline key and found the same getting printed during the hypervisor init. Also I don't see any console related prints in the boot logs of minos-hypervisor. This confirms that the debug console in minos-hypervisor was not initialized.

Let me know what I can try as a next step.

**Comment by lemin9538** [comment]
My RPI3B borad is broken. So I have no board to test. the debug console will creaded by this code path

linux_create_native_vm_resource() ----> create_native_vm_resource_common() --->create_vm_vdev_common() ---> create_vm_vdev_of()

Please help to add some logs in these function to see the reason why the debug console can not be created. After this issue if fixed, you can add below patch to enable the hypervisor debug console.


[rpi3b_debug_console_patch.txt](https://github.com/minosproject/minos/files/5011882/rpi3b_debug_console_patch.txt)

Then rebuild the minos.bin and minos.dtb

# make clean
# make rpi_3_defconfig
# make && make dtb

**Comment by aananthcn** [comment]
Hello Le Min, I haven't tried your "rpi3b_debug_console_patch.txt" patch yet. But based on your first input, I added more debug log prints (see attached patch file [aananths-changes.txt](https://github.com/minosproject/minos/files/5013491/aananths-changes.txt)) and I found that the minos-hypervisor code execution searches for tag "vm0_bdi" from device node "vm0". 

Based on the above finding, I modified the "dtbs/bcm2837-rpi-3-b-plus.dts" file by moving the "vm0_bdi" block into the "vm0" block, then I found that the debug console inside minos-hypervisor got initialised successfully:

```
[       0.091312@00 999] NIC     affinity: 0 1 0 0 0 0 3472328308304117760 0
[       0.098340@00 999] NIC Debug-Aananth: linux_create_native_vm_resource()
[       0.118483@00 999] NIC bcm2836_virqchip_init enter
[       0.123157@00 999] WRN PMD block remap 0x28600000 ---> 0x28600000 @0x200000
[       0.130314@00 999] NIC bcm2836_virqchip_init exit
[       0.474589@00 999] NIC Debug-Aananth: create_vm_vdev_of()
[       0.479601@00 999] NIC Debug-Aananth - node name: timer
[       0.521637@00 999] NIC Debug-Aananth: create_native_vm_resource_common():429
[       0.528318@00 999] NIC Debug-Aananth: name = vm0_bdi, dev_node->name = vm0
[       0.535613@00 999] NIC Debug-Aananth: create_native_vm_resource_common():437
[       0.543082@00 999] NIC Debug-Aananth: create_native_vm_resource_common():444
[       0.550568@00 999] NIC Debug-Aananth: create_native_vm_resource_common():451
[       0.558023@00 999] NIC Debug-Aananth: create_vm_vdev_common()
[       0.564182@00 999] NIC Debug-Aananth: create_vm_vdev_common()
[       0.570334@00 999] NIC Debug-Aananth: create_vm_vdev_of()
[       0.576127@00 999] NIC Debug-Aananth - node name: vm_console_vm0
[       0.582632@00 999] NIC Debug-Aananth: create_dconsole() ENTRY
[       0.588712@00 999] NIC debug_console base: 0x3f20a000
[       0.594145@00 999] NIC Debug-Aananth: create_dconsole() NORMAL EXIT
[       0.606747@00 999] NIC New cmdline: 8250.nr_uarts=1 dwc_otg.lpm_enable=0 net.ifnames=0 earlycon=uart8250,mmio32,0x3f215040 console=ttyS0,115200 cma=64M root=/dev/mmcblk0p2 rw rootfstype=ext4 elevator=deadline fsck.repaird
```

In Linux kernel boot-log, I see mixed results. Good to see that the vmbox hvc console is initialized, but the last line below says that it cannot find the hvc console device! I assume this is the kernel bug that you referred in your earlier comment and it is safe to ignore. Am I right?
```
[    0.008330] Console: colour dummy device 80x25
[    0.012883] register vm debug console
[    0.016592] vmbox hvc console init for backend
[    0.021170] can not find the hvc console device
```

At the end I got the login prompt (guest vm0), I then logged in and found that these device (below) nodes were created. How do I test it? That is, **how to connect to hypervisor's console?**
```
pi@raspberrypi:~$ ls /dev/hvc*
/dev/hvc0  /dev/hvc2  /dev/hvc4  /dev/hvc6
/dev/hvc1  /dev/hvc3  /dev/hvc5  /dev/hvc7
```

**Comment by lemin9538** [comment]
the you need rpi3b_debug_console_patch.txt. you need modify the command line, please refers to this patch.

**Comment by aananthcn** [comment]
I have applied "rpi3b_debug_console_patch.txt + vm0_bdi" patches, reverted all other patches. Reconfigured, built, copied minos.bin & minos.dtb to SD Card as per your instructions, all booted well. But the serial console is connected to the Linux guest. 

How do I attach to minos hypervisor console? (i.e, switching to minos# prompt)

**Comment by lemin9538** [comment]
Ctrl + d - can enter into hypervisor's console

**Comment by aananthcn** [comment]
Pressing Ctrl + d after logging into the Linux guest logs me out. But, after logout, pressing Ctrl + d has no effect. But, if I leave the system on for about 2 hours then I get "Minos #" prompt (I don't know if Linux guest crashed or went to sleep - no logs)

Not sure if I should close this issue and open another for this.

**Comment by lemin9538** [comment]
In hypervisor console enter below command to enter into guest console

# try attach vm0

**Comment by aananthcn** [comment]
The system appeared frozen, when I tried "tty attach vm0", after I get the hypervisor console after about 2 hours. Also, when I get hypervisor console after 2 hours, the HDMI display goes blank and the ssh session also freezes. So, attempting to connect to a frozen/crashed vm0 could also result in hypervisor crash?

**Comment by aananthcn** [comment]
Today the behaviour is different, things are working as expected. After 2 hours of waiting, I got hypervisor console.

```
minos # tty attach vm0
Attach tty: vm0 press any key to active the console
minos # 
```

Then up on pressing ENTER button, I got the following:

```
minos # 
raspberrypi login: 
```

This time I can confirm that there was no crash or freeze behaviour. Then I pressed Ctrl + d, Then I got hypervisor console immediately.

```
raspberrypi login: 
Detach tty: vm0

minos # 
```

Any reason why I am not getting this hypervisor console by pressing Ctrl + d for the first time? Any hints/ideas to try?
